Basic hub tool. Minimalistic event/action addition&deletion (Week 1 issue)

ThompsonTyler opened the pull request:
### Contains
Basic hub tool and textures associated with it.
Includes basic logic for event and action addition/deletion using both a context menu and UIButtons.
Logic is stored in a tree, but currently is not persistent across reloading the same save.

flo commented on src/main/java/org/terasology/scenario/internal/ui/HubToolScreen.java:
```diff
@@ -0,0 +1,204 @@
+            });
+
+            LogicTree tree = new LogicTree(new LogicTreeValue("Scenario", false, assetManager.getAsset("Scenario:scenarioText", Texture.class).get(), true));
+            //LogicTreeValue temp = new LogicTreeValue("Sample Event", true, assetManager.getAsset("Scenario:eventText", Texture.class).get(), false);
```
Commented out code should be removed before a PR.

ThompsonTyler replied:
Thought I caught all of it, guess not! Did a double check with next commit.

flo commented on src/main/java/org/terasology/scenario/internal/ui/LogicItemRenderer.java:
```diff
@@ -0,0 +1,90 @@
+
+import java.util.List;
+
+public class LogicItemRenderer extends AbstractItemRenderer<LogicTreeValue> {
```
This class needs a javadoc explaining why it exists. Looks to me like a copy of an existing one.

ThompsonTyler replied:
This is the only thing I don't explicitly address in next commit; it might look familiar because it's very very similar(might be the same?) as the one from your structure templates. So it doesn't exist already in the engine and therefore would need to be recreated. Eventually more will probably be added for easier user interface that will require a new renderer than what already exists, this is kind of just in preparation for that.

flo commented on src/main/java/org/terasology/scenario/internal/ui/LogicTreeMenuTreeBuilder.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class LogicTreeMenuTreeBuilder {
```
I am also someone who likes to use the builder pattern. However this class is a bit confusing to me. It knows actions by string and also uses them to construct something but does get the logic from outside.

What I don't like about it is the complexity. As anothe developer i have to basically read the complete class in order to understand your other code. Also it is not really a clean separation of concerns/roles. 

e.g. a typical use case for a builder like this could be the separation of the technical construction of something (done by the builder) and the logic of what should be put together (the caller of the builder). In your case the createPrimaryContextMenu contains logic that decised what should be put together which is normally not within the builder. Also it is very strange that a builder has to get data from outside that it then uses to build something or not. 

So in your case I would first have gone for a solution without the builder pattern and then see later on if it makes sense to introduce one.

flo replied:
Also using strings as keys to get something is not so ideal. Especially as they are actually texts that might later get translated. If you want to go for constants you could go for an enum although  you problably
just get a nice syntax by just passing method references as arguments  to addOption

flo commented on src/main/java/org/terasology/scenario/internal/ui/HubToolScreen.java:
```diff
@@ -0,0 +1,204 @@
+import org.terasology.rendering.nui.widgets.UIButton;
+import org.terasology.rendering.nui.widgets.treeView.Tree;
+
+public class HubToolScreen extends CoreScreenLayer {
```
Btw. about the name HUB. I really only know what you mean because I discussed it with you. Others might have no clue what you mean with a HUB tool. Maybe a name like "ScenarioEditor" would be better.

ThompsonTyler replied:
Yeah, I think I will have to address this at some point. Currently it isn't the most prominent issue, so I would prefer to do some thinking a bit first before making the change.

flo commented on src/main/java/org/terasology/scenario/internal/ui/LogicTree.java:
```diff
@@ -0,0 +1,38 @@
+
+import org.terasology.rendering.nui.widgets.treeView.Tree;
+
+public class LogicTree extends Tree<LogicTreeValue> {
```
theoretically it might be possible / a good idea to combine LogicTree and LogicTreeValue into a single class. You just have to see how it turns out later for you. So no need to change it now but just keep in mind that it is an option that we might want to go for for simplicity

ThompsonTyler replied:
Yeah, possibly. Especially with the change to the logicTree being entirely the visual portion basically and the main structure being built from entities. Will have to see how it effects future progress.

ThompsonTyler commented:
Put up a new commit for you Flo. This one is a conversion into the entity structure as we addressed. It is still fairly low detail as everything is still sort of just denoted as string's for the name of the event/actions. It doesn't lend itself perfectly with eventual folders/conditionals yet. But these are aspects that are already explicitly brought up in later week's issues that they are not a huge concern yet. 

I addressed some of your feedback at this point. 

Currently the expansion does not really work, will work on this is near future. The entityTree does persist across saves as well(at least it did when I tested, not sure if I fully tested every case yet though).

flo commented:
Merged it. Good progress so far.

Some points for next PR:
1  I would propose the hub tool gets a component that stores the expanded state of the complete tree by storing the set of expanded nodes (Set<EntityRef>)

2. The event for creating an action should have A Prefab instead of string. Whichshould be used to create the action
